Firestore/Swift/Source/Codable/DocumentReference+WriteEncodable.swift

```diff
@@ -0,0 +1,93 @@
+ * limitations under the License.
+ */
+
+import Foundation
+import FirebaseFirestore
+
+extension DocumentReference {
+  /// Encodes an instance of `Encodable` and overwrites the encoded data
+  /// to the document referred by this `DocumentReference`. If no document exists,
+  /// it is created. If a document already exists, it is overwritten.
+  ///
+  /// See `Firestore.Encoder` for more details about the encoding process.
+  ///
+  /// - Parameters:
+  ///   - value: An instance of `Encodable` to be encoded to a document.
+  ///   - completion: A block to execute once the document has been successfully
+  ///                 written to the server. This block will not be called while
+  ///                 the client is offline, though local changes will be visible
+  ///                 immediately.
+  public func setData<T: Encodable>(from value: T,
+                                    completion: ((Error?) -> Void)? = nil) throws {
+    try setData(from: value, encoder: Firestore.Encoder(), completion: completion)
+  }
+
+  /// Encodes an instance of `Encodable` and overwrites the encoded data
+  /// to the document referred by this `DocumentReference`. If no document exists,
+  /// it is created. If a document already exists, it is overwritten.
+  ///
+  /// See `Firestore.Encoder` for more details about the encoding process.
+  ///
+  /// - Parameters:
+  ///   - value: An instance of `Encodable` to be encoded to a document.
+  ///   - encoder: An encoder instance to use to run the encoding.
+  ///   - completion: A block to execute once the document has been successfully
+  ///                 written to the server. This block will not be called while
+  ///                 the client is offline, though local changes will be visible
+  ///                 immediately.
+  public func setData<T: Encodable>(from value: T,
+                                    encoder: Firestore.Encoder,
+                                    completion: ((Error?) -> Void)? = nil) throws {
+    setData(try encoder.encode(value), completion: completion)
+  }
+
+  /// Updates fields in the document referred to by this `FIRDocumentReference`, using the
+  /// `Encodable` object passed in. If the document does not exist, the update fails and
+  /// the specified completion block receives an error.
+
+  /// - Parameters:
+  ///   - value: An instance of `Encodable` to be used to update the document.
+  ///   - completion: A block to execute when the update is complete. If the update is successful the
+  ///      error parameter will be nil, otherwise it will give an indication of how the update failed.
+  ///      This block will only execute when the client is online and the commit has completed against
+  ///      the server. The completion handler will not be called when the device is offline, though
+  ///      local changes will be visible immediately.
+  public func updateData<T: Encodable>(from value: T,
```

Comment:
This doesn't seem like the right interface.

If you look at the [android docs](https://firebase.google.com/docs/reference/android/com/google/firebase/firestore/DocumentReference.html) we have `update(String field, Object value, Object... moreFieldsAndValues)` (and a `FieldPath` variant as well). This takes essentially pairs of paths and values to update. There's also `update(Map<String, Object> data)`, which takes the same kind of data, just allowing for accumulation of those pairs in a `Map`.

In each of these cases though, the values are *parts* of an object that are mapped onto a larger document.

Consider code where you have a `User` document, where the user has an address field of type `Address`. Update is intended to allow the developer to be able to express changing the Address without updating the whole User:

```java
public class Address {
  // the usual fields
}
public class User {
  // getters/setters, etc
  private Address address;
}

DocumentReference userDoc = db.document(...);
Address newAddress = new Address(...));
userDoc.update("address", newAddress);
```

There's also set with merge, which allows you to only update parts of the user, given a whole user:
```java
DocumentReference userDoc = db.document(...);
User user = new User();
user.setAddress(new Address(...));
userDoc.set(user, SetOptions.mergeFields("address"));  // Just the address field, ignore the rest.
```

The existing Swift [update API](https://firebase.google.com/docs/reference/swift/firebasefirestore/api/reference/Classes/DocumentReference#updatedata_:) takes a `fields` parameter of type `[AnyHashable : Any]`, where the keys can be strings representing paths or actual `FieldPath` objects. That is, without Codable, you can do:

```swift
let userDoc = db.document(...);
let newAddress = ["city": "SF", ...];
userDoc.updateData(["address", newAddress]);
```

The API you have here makes it seem as if all I can update is the whole document (a `User`) which isn't what we want.

If we're going to have a `Codable`-based update API it needs to be able to express something like this:

```swift
let userDoc = db.document(...);
let newAddress = Address(...);
userDoc.updateData("address", newAddress, "name", "John Smith");
```

That is, we need to be able to update separate fields in the document that have different types (address of type Address, and name of type string). This works in Java because we look at each values' type independently via reflection. To make this work in Swift at compile time doesn't seem practical.

So I think really what's needed here is overloads for `setData` that take your Codable object and also `merge` and `mergeFields` like we do for [non-codable data](https://firebase.google.com/docs/reference/swift/firebasefirestore/api/reference/Classes/DocumentReference#writing-data).

The way to make `updateData` work would would be to provide overloads with independent template types, something like:

```swift
public func updateData<T: Encodable>(_ field: string, _ value: T);

public func updateData<T: Encodable, U: Encodable>(_ field1: string, _ value1: T, _ field2: string, _ value2: U);

public func updateData<T: Encodable, U: Encodable, V: Encodable>(_ field1: string, _ value1: T, _ field2: string, _ value2: U, _ field3: string, _ value3: V);
```
This is really gross though so I think it's unlikely we should do this.

Reply:
Thanks for the detailed explanation. I felt like `updateData` implemented that way is overlapping with `setData:merge`, maybe I should've discussed this with you before I proceed.

Anyways. Did what you suggested.